File: personalizatio-sdk/src/main/java/com/personalizatio/stories/views/storyItem/StoryItemView.java
package com.personalizatio.stories.views.storyItem;

import android.annotation.SuppressLint;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.Log;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import androidx.media3.ui.PlayerView;
import androidx.recyclerview.widget.RecyclerView;
import androidx.transition.ChangeBounds;
import androidx.transition.Transition;
import androidx.transition.TransitionManager;
import androidx.transition.TransitionSet;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.material.shape.CornerFamily;
import com.google.android.material.shape.MaterialShapeDrawable;
import com.google.android.material.shape.ShapeAppearanceModel;
import com.google.common.base.Strings;
import com.personalizatio.Product;
import com.personalizatio.R;
import com.personalizatio.SDK;
import com.personalizatio.stories.models.elements.ButtonElement;
import com.personalizatio.stories.models.elements.HeaderElement;
import com.personalizatio.stories.models.elements.ProductElement;
import com.personalizatio.stories.models.elements.ProductsElement;
import com.personalizatio.stories.models.elements.TextBlockElement;
import com.personalizatio.stories.viewAdapters.ProductsAdapter;
import com.personalizatio.stories.models.elements.Element;
import com.personalizatio.stories.models.Slide;
import com.personalizatio.stories.views.StoriesView;
import com.personalizatio.utils.ViewUtils;

@SuppressLint("ViewConstructor")
final public class StoryItemView extends ConstraintLayout {

	public interface OnPageListener {
		void onPrev();

		void onNext();

		void onPrepared(int position);

		void onLocked(boolean lock);
	}

	private ImageView image;
	private PlayerView video;
	private ConstraintLayout product;
	private LinearLayout productPriceBox;
	private LinearLayout productDiscountBox;
	private TextView productBrand;
	private TextView productName;
	private TextView productOldPrice;
	private TextView productPrice;
	private TextView productDiscount;
	private TextView promocodeText;
	private ImageView productImage;

	private OnPageListener pageListener;
	private View prev;
	private View next;
	private final StoriesView storiesView;

	//Элементы управления
	private Button button;
	private FrameLayout textBlocksLayout;
	public ImageButton reload;
	public View reload_layout;
	private TextView reloadText;
	private ConstraintLayout header;
	private TextView titleTextView;
	private TextView subtitleTextView;
	private CardView titleCardView;
	private ImageView titleIconImageView;
	private Button buttonProducts;
	private ViewGroup elementsLayout;
	private RecyclerView products;
	private ProductsAdapter productsAdapter;

	private int viewHeight;
	private int viewTopOffset;

	private final int ELEMENTS_LAYOUT_ANIMATION_DELAY = 100;
	private final int ELEMENTS_LAYOUT_ANIMATION_DURATION = 200;

	public StoryItemView(StoriesView storiesView) {
		super(storiesView.getContext());
		this.storiesView = storiesView;

		inflate(getContext(), R.layout.story_item, this);
		setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));

		initViews();
		setupViews();
	}

	private void initViews() {
		image = findViewById(android.R.id.background);
		video = findViewById(android.R.id.widget_frame);

		//Переключение слайда
		prev = findViewById(R.id.reverse);
		next = findViewById(R.id.skip);

		//Элементы управления
		button = findViewById(android.R.id.button1);
		textBlocksLayout = findViewById(R.id.text_blocks_layout);
		reload = findViewById(R.id.reload);
		reload_layout = findViewById(R.id.reload_layout);
		reloadText = findViewById(R.id.reload_text);
		header = findViewById(R.id.header);
		titleTextView = findViewById(R.id.title_textView);
		subtitleTextView = findViewById(R.id.subtitle_textView);
		titleIconImageView = findViewById(R.id.title_imageView);
		titleCardView = findViewById(R.id.titleCardView);
		buttonProducts = findViewById(android.R.id.button2);
		elementsLayout = findViewById(R.id.elements_layout);

		products = findViewById(android.R.id.list);

		//Товарный слайд
		product = findViewById(R.id.product);

		productPriceBox = findViewById(R.id.product_price_box);
		productBrand = findViewById(R.id.product_brand);
		productName = findViewById(R.id.product_name);
		productOldPrice = findViewById(R.id.product_oldprice);
		productPrice = findViewById(R.id.product_price);
		productDiscount = findViewById(R.id.product_discount);
		productDiscountBox = findViewById(R.id.product_discount_box);
		productImage = findViewById(R.id.product_image);
		promocodeText = findViewById(R.id.promocode_text);
	}

	private void setupViews() {
		image.setVisibility(View.GONE);
		video.setVisibility(View.GONE);

		//Переключение слайда
		prev.setOnClickListener((View) -> {
			if( pageListener != null ) {
				pageListener.onPrev();
			}
		});
		next.setOnClickListener((View) -> {
			if( pageListener != null ) {
				pageListener.onNext();
			}
		});

		productsAdapter = new ProductsAdapter(storiesView);
		products.setAdapter(productsAdapter);

		product.setVisibility(GONE);
	}

	/**
	 * Слушатель переключения слайда
	 * @param listener OnPageListener
	 */
	public void setOnPageListener(OnPageListener listener) {
		pageListener = listener;
	}

	public void setOnTouchListener(OnTouchListener listener) {
		prev.setOnTouchListener(listener);
		next.setOnTouchListener(listener);
		super.setOnTouchListener((v, event) -> true);
	}

	public void setViewSize(int height, int topOffset) {
		viewHeight = height;
		viewTopOffset = topOffset;
	}

	/**
	 * Обновляет данные слайда
	 *
	 * @param slide Story.Slide
	 */
	@SuppressLint("ResourceAsColor")
    public void update(Slide slide, int position, String code, int story_id) {
		slide.setPrepared(false);

		setBackgroundColor(ViewUtils.getColor(getContext(), slide.getBackgroundColor(), android.R.color.black));

		video.setVisibility(GONE);

		setupReloadView(slide, position, code, story_id);

		setButtonOnClickListener(slide, code, story_id);

		//Загуражем картинку
		if( slide.getType().equals("image") ) {
			loadImage(slide.getBackground(), new RequestListener<>() {
				@Override
				public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
					reload_layout.setVisibility(VISIBLE);
					return false;
				}

				@Override
				public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
					onPreparedSlide(slide, position);
					return false;
				}
			});
		}

		//Загружаем видео
		if (slide.getType().equals("video")) {
			video.setVisibility(VISIBLE);

			//Загружаем превью
			var preview = slide.getPreview();
			if (preview != null) {
				loadImage(preview, new RequestListener<>() {
					@Override
					public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
						return false;
					}

					@Override
					public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
						return false;
					}
				});
			}
		}

		//Обновляем элементы слайда
		setupElements(slide, code, story_id, position);
	}

	public ImageView getImage() {
		return image;
	}

	public PlayerView getVideo() {
		return video;
	}

	private void onPreparedSlide(Slide slide, int position) {
		slide.setPrepared(true);
		reload_layout.setVisibility(GONE);
		elementsLayout.setVisibility(VISIBLE);
		if (!buttonProducts.isActivated() && pageListener != null) {
			pageListener.onPrepared(position);
		}
	}

	private void setButtonOnClickListener(Slide slide, String code, int storyId) {
		button.setOnClickListener(view -> {
			try {
				Product product = null;
				String link = null;
				//Сначала ищем элемент с товаром
				for (var element : slide.getElements()) {
					if (element instanceof ProductElement productElement) {
						product = productElement.getItem();
					} else if (element instanceof ButtonElement buttonElement) {
						link = buttonElement.getLink();
					}
				}
				Log.d(SDK.TAG, "open link: " + link + (product != null ? " with product: `" + product.getId() + "`" : ""));
				//Вызываем колбек клика
				var clickListener = storiesView.getClickListener();
				if( clickListener == null
						|| product == null && clickListener.onClick(link)
						|| product != null && clickListener.onClick(product) ) {
					getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(link)));
				}
				SDK.track_story("click", code, storyId, slide.getId());
			} catch(ActivityNotFoundException | NullPointerException e) {
				Log.e(SDK.TAG, e.getMessage(), e);
				Toast.makeText(getContext(), "Unknown error", Toast.LENGTH_SHORT).show();
			}
		});
	}

	/**
	 * Загружает изображение слайда
	 *
	 * @param url      String
	 * @param listener RequestListener<Drawable>
	 */
	private void loadImage(String url, RequestListener<Drawable> listener) {
		image.setVisibility(View.VISIBLE);
		Glide.with(getContext()).load(url).listener(listener).into(image);
	}

	public void updateProduct(ProductElement element, RequestListener<Drawable> listener) {
		var item = element.getItem();

		product.setVisibility(VISIBLE);

		Glide.with(getContext()).load(item.getImage()).listener(listener).override(Target.SIZE_ORIGINAL).into(productImage);

		setupDefaultTextView(productBrand, Strings.isNullOrEmpty(item.getBrand()), item.getBrand());
		setupDefaultTextView(productName, item.getName());
		setupDefaultTextView(productPrice, item.getPrice());
		setupDefaultTextView(productOldPrice, Strings.isNullOrEmpty(item.getOldPrice()), item.getOldPrice());
		setupDefaultTextView(promocodeText, element.getTitle() != null && !Strings.isNullOrEmpty(item.getPromocode()), element.getTitle());
		productOldPrice.setPaintFlags(productOldPrice.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
		productDiscountBox.setVisibility(Strings.isNullOrEmpty(item.getDiscountPercent()) && Strings.isNullOrEmpty(item.getPromocode()) ? GONE : VISIBLE);

		//Указываем закругления
		float radius = getResources().getDimension(R.dimen.product_price_box_radius);
		ShapeAppearanceModel shape_box = new ShapeAppearanceModel().toBuilder().setAllCorners(CornerFamily.ROUNDED, radius).build();
		MaterialShapeDrawable shape_box_drawable = new MaterialShapeDrawable(shape_box);
		ShapeAppearanceModel shape_discount = new ShapeAppearanceModel().toBuilder()
				.setTopRightCorner(CornerFamily.ROUNDED, radius)
				.setBottomRightCorner(CornerFamily.ROUNDED, radius)
				.build();
		MaterialShapeDrawable shape_discount_drawable = new MaterialShapeDrawable(shape_discount);

		//Заполняем
		shape_box_drawable.setFillColor(ContextCompat.getColorStateList(getContext(), android.R.color.white));
		ViewCompat.setBackground(productPriceBox, shape_box_drawable);

		//Блок скидки или промокода
		var productDiscountText = "";
		int shapeDiscountColor = 0;
		if (!Strings.isNullOrEmpty(item.getPromocode())) {
			productDiscountText = item.getPromocode();
			shapeDiscountColor = R.color.product_promocode_color;
			productPrice.setText(item.getPriceWithPromocode());
		} else if (!Strings.isNullOrEmpty(item.getDiscountPercent())) {
			productDiscountText = "-" + item.getDiscountPercent() + "%";
			shapeDiscountColor = R.color.product_discount_color;
		}
		setupDefaultTextView(productDiscount, productDiscountText);
		if(shapeDiscountColor != 0) {
			shape_discount_drawable.setFillColor(ContextCompat.getColorStateList(getContext(), shapeDiscountColor));
		}

		ViewCompat.setBackground(productDiscountBox, shape_discount_drawable);
	}

	public void release() {
		video.setPlayer(null);
	}

	public void setHeadingVisibility(int visibility) {
		elementsLayout.animate().alpha(visibility == GONE ? 0 : 1).setStartDelay(visibility == GONE ? ELEMENTS_LAYOUT_ANIMATION_DELAY : 0).setDuration(ELEMENTS_LAYOUT_ANIMATION_DURATION);
	}

	private void updateHeader(HeaderElement element, String slide_id, String code, int story_id) {
		header.setVisibility(VISIBLE);
		header.setOnTouchListener((View v, MotionEvent event) -> {
			if( event.getAction() == MotionEvent.ACTION_UP ) {
				getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(element.getLink())));
				SDK.track_story("click", code, story_id, slide_id);
			}
			return true;
		});

		if( element.getIcon() != null ) {
			titleCardView.setVisibility(VISIBLE);
			if( getContext() == null ) return;
			Glide.with(getContext()).load(element.getIcon()).into(titleIconImageView);
		} else {
			titleCardView.setVisibility(GONE);
		}

		var title = element.getTitle();
		setupDefaultTextView(titleTextView, title != null, title);

		var subtitle = element.getSubtitle();
		setupDefaultTextView(subtitleTextView, subtitle != null, subtitle);
	}

	private void setupElements(Slide slide, String code, int story_id, int position) {
		//Скрываем все элементы
		header.setVisibility(GONE);
		button.setVisibility(GONE);
		buttonProducts.setVisibility(GONE);
		products.setVisibility(GONE);

		//Отображаем необходимые элементы
		for (Element element : slide.getElements()) {
			if (element instanceof HeaderElement headerElement) {
				updateHeader(headerElement, slide.getId(), code, story_id);
			} else if (element instanceof TextBlockElement textBlockElement) {
				updateTextBlock(textBlockElement);
			} else if (element instanceof ButtonElement buttonElement) {
				updateButton(buttonElement);
			}
			else if (element instanceof ProductsElement productsElement) {
				updateProducts(productsElement, slide.getId(), story_id);
			} else if (element instanceof ProductElement productElement) {
				updateProduct(productElement, new RequestListener<>() {
					@Override
					public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
						return false;
					}

					@Override
					public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
						slide.setPrepared(true);
						if( pageListener != null ) {
							pageListener.onPrepared(position);
						}
						return false;
					}
				});
			}
		}
	}

	private void updateProducts(ProductsElement element, String slideId, int storyId) {
		productsAdapter.setProducts(element.getProducts(), storyId, slideId);
		setupTextView(buttonProducts, true, element.getLabelShow(), storiesView.getSettings().products_button_font_family);
		buttonProducts.setOnClickListener(view -> {
			buttonProducts.setActivated(!buttonProducts.isActivated());
			buttonProducts.setText(buttonProducts.isActivated() ? element.getLabelHide() : element.getLabelShow());

			//Анимация появления товаров
			ConstraintSet set = new ConstraintSet();
			set.clone((ConstraintLayout) elementsLayout);

			Transition transition = new ChangeBounds();
			transition.addTarget(buttonProducts.getId());
			transition.addTarget(button.getId());

			Transition transition2 = new androidx.transition.Slide(Gravity.BOTTOM);
			transition2.addTarget(products.getId());

			TransitionSet transitions = new TransitionSet();
			transitions.addTransition(transition);
			transitions.addTransition(transition2);

			TransitionManager.beginDelayedTransition(elementsLayout, transitions);
			products.setVisibility(buttonProducts.isActivated() ? VISIBLE : GONE);
			//--->

			if( pageListener != null ) {
				pageListener.onLocked(buttonProducts.isActivated());
			}
		});
		if( pageListener != null ) {
			pageListener.onLocked(buttonProducts.isActivated());
		}
	}

	@SuppressLint("ResourceAsColor")
	private void updateButton(ButtonElement element) {
		button.setVisibility(VISIBLE);
		button.setText(element.getTitle());

		ViewUtils.setBackgroundColor(getContext(), button, element.getBackground(), R.color.primary);
		ViewUtils.setTextColor(getContext(), button, element.getColor(), R.color.white);

		button.setTypeface(Typeface.create(storiesView.getSettings().button_font_family, element.getTextBold() ? Typeface.BOLD : Typeface.NORMAL));
	}

	private void updateTextBlock(TextBlockElement element) {
		var textBlockView = new TextBlockView(getContext());

		textBlockView.updateView(element, viewHeight, viewTopOffset);

		textBlocksLayout.addView(textBlockView);
	}

	private void setupReloadView(Slide slide, int position, String code, int storyId) {
		reload_layout.setVisibility(GONE);

		var settings = storiesView.getSettings();
		setupTextView(reloadText, settings.failed_load_text, settings.failed_load_font_family);
		reloadText.setTextSize(storiesView.getSettings().failed_load_size);
		reloadText.setTextColor(Color.parseColor(storiesView.getSettings().failed_load_color));

		reload.setOnClickListener((View) -> update(slide, position, code, storyId));
	}

	private void setupDefaultTextView(TextView textView, boolean visibility, String text) {
		setupTextView(textView, visibility, text, storiesView.getSettings().font_family);
	}

	private void setupTextView(TextView textView, boolean visibility, String text, Typeface typeface) {
		textView.setVisibility(visibility ? VISIBLE : GONE);

		if (visibility) {
			setupTextView(textView, text, typeface);
		}
	}

	private void setupDefaultTextView(TextView textView, String text) {
		setupTextView(textView, text, storiesView.getSettings().font_family);
	}

	private void setupTextView(TextView textView, String text, Typeface typeface) {
		textView.setText(text);
		textView.setTypeface(typeface);
	}
}

File: personalizatio-sdk/src/main/java/com/personalizatio/stories/StoryItemView.kt
package com.personalizatio.stories

import android.annotation.TargetApi

internal class StoryItemView : ConstraintLayout {
    interface OnPageListener {
        fun onPrev()

        fun onNext()

        fun onPrepared(position: Int)

        fun onLocked(lock: Boolean)
    }

    var image: ImageView? = null
    var video: PlayerView? = null
    private var product: ConstraintLayout? = null
    private var product_price_box: LinearLayout? = null
    private var product_discount_box: LinearLayout? = null
    private var product_brand: TextView? = null
    private var product_name: TextView? = null
    private var product_oldprice: TextView? = null
    private var product_price: TextView? = null
    private var product_discount: TextView? = null
    private var promocode_text: TextView? = null
    private var product_image: ImageView? = null

    private var page_listener: OnPageListener? = null
    private var prev: View? = null
    private var next: View? = null
    private var stories_view: StoriesView? = null

    //Элементы управления
    private var button: Button? = null
    var reload: ImageButton? = null
    var reload_layout: View? = null
    private var reload_text: TextView? = null
    private var header: ConstraintLayout? = null
    private var titleTextView: TextView? = null
    private var subtitleTextView: TextView? = null
    private var titleCardView: CardView? = null
    private var titleIconImageView: ImageView? = null
    private var button_products: Button? = null
    private var elements_layout: ViewGroup? = null
    private var products: RecyclerView? = null
    private var products_adapter: ProductsAdapter? = null

    constructor(@NonNull context: Context?) : super(context)

    constructor(@NonNull context: Context?, @Nullable attrs: AttributeSet?) : super(context, attrs)

    constructor(@NonNull context: Context?, @Nullable attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        @NonNull context: Context?,
        @Nullable attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    @Override
    protected fun onFinishInflate() {
        super.onFinishInflate()
        init()
    }

    private fun init() {
        image = findViewById(android.R.id.background)
        video = findViewById(android.R.id.widget_frame)
        image.setVisibility(View.GONE)
        video.setVisibility(View.GONE)

        //Переключение слайда
        prev = findViewById(R.id.reverse)
        next = findViewById(R.id.skip)
        prev.setOnClickListener { View ->
            if (page_listener != null) {
                page_listener.onPrev()
            }
        }
        next.setOnClickListener { View ->
            if (page_listener != null) {
                page_listener.onNext()
            }
        }

        //Элементы управления
        button = findViewById(android.R.id.button1)
        reload = findViewById(R.id.reload)
        reload_layout = findViewById(R.id.reload_layout)
        reload_text = findViewById(R.id.reload_text)
        header = findViewById(R.id.header)
        titleTextView = findViewById(R.id.title_textView)
        subtitleTextView = findViewById(R.id.subtitle_textView)
        titleIconImageView = findViewById(R.id.title_imageView)
        titleCardView = findViewById(R.id.titleCardView)
        button_products = findViewById(android.R.id.button2)
        elements_layout = findViewById(R.id.elements_layout)

        products_adapter = ProductsAdapter()
        products = findViewById(android.R.id.list)
        products.setAdapter(products_adapter)

        //Товарный слайд
        product = findViewById(R.id.product)
        product.setVisibility(GONE)

        product_price_box = findViewById(R.id.product_price_box)
        product_brand = findViewById(R.id.product_brand)
        product_name = findViewById(R.id.product_name)
        product_oldprice = findViewById(R.id.product_oldprice)
        product_price = findViewById(R.id.product_price)
        product_discount = findViewById(R.id.product_discount)
        product_discount_box = findViewById(R.id.product_discount_box)
        product_image = findViewById(R.id.product_image)
        promocode_text = findViewById(R.id.promocode_text)
    }

    /**
     * Слушатель переключения слайда
     * @param listener OnPageListener
     */
    fun setOnPageListener(listener: OnPageListener?) {
        page_listener = listener
    }

    /**
     * Слушатель клика по ссылке
     * @param view StoriesView
     */
    fun setStoriesView(view: StoriesView?) {
        stories_view = view
        products_adapter.setStoriesView(view)
    }

    fun setOnTouchListener(listener: OnTouchListener?) {
        prev.setOnTouchListener(listener)
        next.setOnTouchListener(listener)
        super.setOnTouchListener { v, event -> true }
    }

    /**
     * Обновляет данные слайда
     *
     * @param slide Story.Slide
     */
    fun update(slide: Story.Slide?, position: Int, code: String?, story_id: Int) {
        slide.prepared = false
        setBackgroundColor(
            if (slide.background_color == null) getContext().getResources()
                .getColor(android.R.color.black) else Color.parseColor(slide.background_color)
        )
        video.setVisibility(GONE)
        reload_layout.setVisibility(GONE)
        reload_text.setTypeface(stories_view.settings.failed_load_font_family)
        reload_text.setTextSize(stories_view.settings.failed_load_size)
        reload_text.setText(stories_view.settings.failed_load_text)
        reload_text.setTextColor(Color.parseColor(stories_view.settings.failed_load_color))

        reload.setOnClickListener { View ->
            update(slide, position, code, story_id)
        }

        //Вызываем клик по кнопке
        button.setOnClickListener { view ->
            try {
                var product: Product? = null
                var link: String? = null
                //Сначала ищем элемент с товаром
                for (element in slide.elements) {
                    when (element.type) {
                        "product" -> product = element.item
                        "button" -> link = element.link
                    }
                }
                Log.d(
                    SDK.TAG,
                    "open link: " + link + (if (product != null) (" with product: `" + product.id).toString() + "`" else "")
                )
                //Вызываем колбек клика
                if (stories_view.click_listener == null || product == null && stories_view.click_listener.onClick(link) || product != null && stories_view.click_listener.onClick(
                        product
                    )
                ) {
                    getContext().startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(link)))
                }
                SDK.track_story("click", code, story_id, slide.id)
            } catch (e: ActivityNotFoundException) {
                Log.e(SDK.TAG, e.getMessage(), e)
                Toast.makeText(getContext(), "Unknown error", Toast.LENGTH_SHORT).show()
            } catch (e: NullPointerException) {
                Log.e(SDK.TAG, e.getMessage(), e)
                Toast.makeText(getContext(), "Unknown error", Toast.LENGTH_SHORT).show()
            }
        }

        //Загуражем картинку
        if (slide.type.equals("image")) {
            loadImage(slide.background, object : RequestListener() {
                @Override
                fun onLoadFailed(
                    @Nullable e: GlideException?,
                    model: Object?,
                    target: Target<Drawable?>?,
                    isFirstResource: Boolean
                ): Boolean {
                    reload_layout.setVisibility(VISIBLE)
                    return false
                }

                @Override
                fun onResourceReady(
                    resource: Drawable?,
                    model: Object?,
                    target: Target<Drawable?>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    onPreparedSlide(slide, position)
                    return false
                }
            })
        }

        //Загружаем видео
        if (slide.type.equals("video")) {
            video.setVisibility(VISIBLE)

            //Загружаем превью
            if (slide.preview != null) {
                loadImage(slide.preview, object : RequestListener() {
                    @Override
                    fun onLoadFailed(
                        @Nullable e: GlideException?,
                        model: Object?,
                        target: Target<Drawable?>?,
                        isFirstResource: Boolean
                    ): Boolean {
                        return false
                    }

                    @Override
                    fun onResourceReady(
                        resource: Drawable?,
                        model: Object?,
                        target: Target<Drawable?>?,
                        dataSource: DataSource?,
                        isFirstResource: Boolean
                    ): Boolean {
                        return false
                    }
                })
            }
        }

        //Обновляем элементы слайда
        updateElements(slide, code, story_id, position)
    }

    private fun onPreparedSlide(slide: Story.Slide?, position: Int) {
        slide.prepared = true
        reload_layout.setVisibility(GONE)
        elements_layout.setVisibility(VISIBLE)
        if (!button_products.isActivated() && page_listener != null) {
            page_listener.onPrepared(position)
        }
    }

    /**
     * Загружает изображение слайда
     *
     * @param url      String
     * @param listener RequestListener<Drawable>
    </Drawable> */
    private fun loadImage(url: String?, listener: RequestListener<Drawable?>?) {
        image.setVisibility(View.VISIBLE)
        Glide.with(getContext()).load(url).listener(listener).into(image)
    }

    fun updateProduct(element: Story.Slide.Element?, listener: RequestListener<Drawable?>?) {
        product.setVisibility(VISIBLE)
        product_brand.setVisibility(if (element.item.brand == null) GONE else VISIBLE)
        product_brand.setText(element.item.brand)
        product_brand.setTypeface(stories_view.settings.font_family)
        product_name.setText(element.item.name)
        product_name.setTypeface(stories_view.settings.font_family)
        Glide.with(getContext()).load(element.item.image).listener(listener).override(Target.SIZE_ORIGINAL)
            .into(product_image)
        product_price.setText(element.item.price)
        product_price.setTypeface(stories_view.settings.font_family)
        product_oldprice.setVisibility(if (element.item.oldprice == null) GONE else VISIBLE)
        product_oldprice.setText(element.item.oldprice)
        product_oldprice.setPaintFlags(product_oldprice.getPaintFlags() or Paint.STRIKE_THRU_TEXT_FLAG)
        product_oldprice.setTypeface(stories_view.settings.font_family)
        product_discount_box.setVisibility(if (element.item.discount_percent == null && element.item.promocode == null) GONE else VISIBLE)
        promocode_text.setText(element.title)
        promocode_text.setVisibility(if (element.title == null || element.item.promocode == null) GONE else VISIBLE)
        promocode_text.setTypeface(stories_view.settings.font_family)

        //Указываем закругления
        val radius: Float = getResources().getDimension(R.dimen.product_price_box_radius)
        val shape_box: ShapeAppearanceModel =
            ShapeAppearanceModel().toBuilder().setAllCorners(CornerFamily.ROUNDED, radius).build()
        val shape_box_drawable: MaterialShapeDrawable = MaterialShapeDrawable(shape_box)
        val shape_discount: ShapeAppearanceModel = ShapeAppearanceModel().toBuilder()
            .setTopRightCorner(CornerFamily.ROUNDED, radius)
            .setBottomRightCorner(CornerFamily.ROUNDED, radius)
            .build()
        val shape_discount_drawable: MaterialShapeDrawable = MaterialShapeDrawable(shape_discount)

        //Заполняем
        shape_box_drawable.setFillColor(ContextCompat.getColorStateList(getContext(), android.R.color.white))
        ViewCompat.setBackground(product_price_box, shape_box_drawable)

        //Блок скидки или промокода
        product_discount.setTypeface(stories_view.settings.font_family)
        if (element.item.promocode != null) {
            product_discount.setText(element.item.promocode)
            product_price.setText(element.item.price_with_promocode)
            shape_discount_drawable.setFillColor(
                ContextCompat.getColorStateList(
                    getContext(),
                    R.color.product_promocode_color
                )
            )
        } else if (element.item.discount_percent != null) {
            product_discount.setText(("-" + element.item.discount_percent).toString() + "%")
            shape_discount_drawable.setFillColor(
                ContextCompat.getColorStateList(
                    getContext(),
                    R.color.product_discount_color
                )
            )
        }

        ViewCompat.setBackground(product_discount_box, shape_discount_drawable)
    }

    fun release() {
        video.setPlayer(null)
    }

    fun setHeadingVisibility(visibility: Int) {
        elements_layout.animate().alpha(if (visibility == GONE) 0 else 1)
            .setStartDelay(if (visibility == GONE) 100 else 0).setDuration(200)
    }

    private fun updateHeader(element: Story.Slide.Element?, slide_id: String?, code: String?, story_id: Int) {
        if (element.type.equals("header")) {
            header.setVisibility(VISIBLE)
            header.setOnTouchListener { v: View?, event: MotionEvent? ->
                if (event.getAction() === MotionEvent.ACTION_UP) {
                    getContext().startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(element.link)))
                    SDK.track_story("click", code, story_id, slide_id)
                }
                true
            }

            if (element.icon != null) {
                titleCardView.setVisibility(VISIBLE)
                if (getContext() == null) return
                Glide.with(getContext()).load(element.icon).into(titleIconImageView)
            } else {
                titleCardView.setVisibility(GONE)
            }

            titleTextView.setTypeface(stories_view.settings.font_family)
            if (element.title != null) {
                titleTextView.setVisibility(VISIBLE)
                titleTextView.setText(element.title)
            } else {
                titleTextView.setVisibility(GONE)
            }

            subtitleTextView.setTypeface(stories_view.settings.font_family)
            if (element.subtitle != null) {
                subtitleTextView.setVisibility(VISIBLE)
                subtitleTextView.setText(element.subtitle)
            } else {
                subtitleTextView.setVisibility(GONE)
            }
        }
    }

    private fun updateElements(slide: Story.Slide?, code: String?, story_id: Int, position: Int) {
        //Скрываем все элементы

        header.setVisibility(GONE)
        button.setVisibility(GONE)
        button_products.setVisibility(GONE)
        products.setVisibility(GONE)

        //Отображаем необходимые элементы
        for (element in slide.elements) {
            when (element.type) {
                "header" -> updateHeader(element, slide.id, code, story_id)
                "button" -> {
                    button.setVisibility(VISIBLE)
                    button.setText(element.title)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        button.setBackgroundTintList(
                            ColorStateList.valueOf(
                                if (element.background == null) button.getContext().getResources()
                                    .getColor(R.color.primary) else Color.parseColor(element.background)
                            )
                        )
                    } else {
                        button.setBackgroundColor(
                            if (element.background == null) button.getContext().getResources()
                                .getColor(R.color.primary) else Color.parseColor(element.background)
                        )
                    }
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        button.setTextColor(
                            ColorStateList.valueOf(
                                if (element.color == null) button.getContext().getResources()
                                    .getColor(R.color.white) else Color.parseColor(element.color)
                            )
                        )
                    } else {
                        button.setTextColor(
                            if (element.color == null) button.getContext().getResources()
                                .getColor(R.color.white) else Color.parseColor(element.color)
                        )
                    }
                    button.setTypeface(
                        Typeface.create(
                            stories_view.settings.button_font_family,
                            if (element.text_bold) Typeface.BOLD else Typeface.NORMAL
                        )
                    )
                }

                "products" -> {
                    products_adapter.setProducts(element.products, story_id, slide.id)
                    button_products.setVisibility(VISIBLE)
                    button_products.setText(element.label_show)
                    button_products.setTypeface(stories_view.settings.products_button_font_family)
                    button_products.setOnClickListener { view ->
                        button_products.setActivated(!button_products.isActivated())
                        button_products.setText(if (button_products.isActivated()) element.label_hide else element.label_show)

                        //Анимация появления товаров
                        val set: ConstraintSet = ConstraintSet()
                        set.clone(elements_layout as ConstraintLayout?)

                        val transition: Transition = ChangeBounds()
                        transition.addTarget(button_products.getId())
                        transition.addTarget(button.getId())

                        val transition2: Transition = Slide(Gravity.BOTTOM)
                        transition2.addTarget(products.getId())

                        val transitions: TransitionSet = TransitionSet()
                        transitions.addTransition(transition)
                        transitions.addTransition(transition2)

                        TransitionManager.beginDelayedTransition(elements_layout, transitions)
                        products.setVisibility(if (button_products.isActivated()) VISIBLE else GONE)

                        //--->
                        if (page_listener != null) {
                            page_listener.onLocked(button_products.isActivated())
                        }
                    }
                    if (page_listener != null) {
                        page_listener.onLocked(button_products.isActivated())
                    }
                }

                "product" -> updateProduct(element, object : RequestListener() {
                    @Override
                    fun onLoadFailed(
                        @Nullable e: GlideException?,
                        model: Object?,
                        target: Target<Drawable?>?,
                        isFirstResource: Boolean
                    ): Boolean {
                        return false
                    }

                    @Override
                    fun onResourceReady(
                        resource: Drawable?,
                        model: Object?,
                        target: Target<Drawable?>?,
                        dataSource: DataSource?,
                        isFirstResource: Boolean
                    ): Boolean {
                        slide.prepared = true
                        if (page_listener != null) {
                            page_listener.onPrepared(position)
                        }
                        return false
                    }
                })
            }
        }
    }
}
